dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels2.java

```diff
@@ -159,11 +161,29 @@ public String toString() {
                     result = DialogueFutures.transform(
                             executeWithStickyToken(request, endpointChannel), successTransformer);
                     result = DialogueFutures.catchingAllAsync(result, failureTransformer);
-                    callInFlight = Futures.nonCancellationPropagating(result);
+                    callInFlight = result;
                 } else {
-                    result = callInFlightSnapshot;
-                    result = DialogueFutures.transformAsync(result, _input -> execute(request, endpointChannel));
-                    result = DialogueFutures.catchingAllAsync(result, _throwable -> execute(request, endpointChannel));
+                    // Each subsequent (parallel) call may be independently cancelled, that cancellation
+                    // must not leak to other pending calls.
+                    result = Futures.nonCancellationPropagating(callInFlightSnapshot);
+                    // State must be tracked to prevent duplicate requests when the original in-flight request
+                    // completes, triggering waiting requests, which may fail via an exception and get
+                    // unintentionally retried by 'catchingAllAsync'.
+                    AtomicBoolean handled = new AtomicBoolean();
+                    result = DialogueFutures.transformAsync(result, response -> {
```

Comment:
What would be wrong with this impl:

```
                    result = Futures.nonCancellationPropagating(callInFlightSnapshot);
                    SettableFuture<Response> response = SettableFuture.create();
                    DialogueFutures.addDirectCallback(result, new FutureCallback<>() {
                        @Override
                        public void onSuccess(Response result) {
                            response.setFuture(execute(request, endpointChannel));
                        }

                        @Override
                        public void onFailure(Throwable _ignored) {
                            response.setFuture(execute(request, endpointChannel));
                        }
                    });
```

Reply:
`setFuture` can leak responses silently, so we have to wire it up ourselves. I've pushed something similar which I think handles that for us, with a great many additional test cases.

I caught another bug in which `catchingAllAsync` didn't catch cancellation (by design) which left this object in a bad state...

Reply:
> setFuture can leak responses silently, so we have to wire it up ourselves. I've pushed something similar which I think handles that for us, with a great many additional test cases.

wait what? Is that a bug or by design?

Reply:
By design, guava future.cancel always moves a future to the cancelled state, however it can be invoked while a transformation is in progress:

consumerFuture.cancel -> hmm <- producerFuture.complete

DialogueFutures methods return false from cancel if there's a transformation in progress, so we can add a listener to consume the resulting future.